### ISSUE TITLE
High CPU system time on task creation contention

### ISSUE THREAD

**Issue by Lastique** [01org/tbb#105]
We have a problem with TBB generating very high CPU overhead in kernel (i.e. system time), when TBB tasks are generated by many threads. For example, our application has 256 IO threads, each periodically producing some data, which has to be processed in a number of consumers (let's say, also 256 per IO thread). This processing is parallelized by using `tbb::parallel_for`, so that when the number of consumers is high and the number of producers is low, the processing doesn't get bottlenecked by the IO thread performance. On a 4-core CPU with Hyperthreading this results in ~95% total CPU utilization, with about 25%-30% of that load being the kernel time, as reported by htop.

From inspecting TBB code and profiling with Oprofile, it looks like the problem is caused by excessive spinning inside TBB when tasks are created by IO threads or stolen by the TBB worker threads. This causes lots of context switches with no real work done. If my hypotheses is correct, I would like to ask to modify TBB implementation so that the excessive spinning is avoided by blocking the threads e.g. on a futex after a few spins.

I've put together a standalone test that resembles our application behavior. Its CPU load is somewhat different from our application (htop shows ~95% total load with about 70% of the system load), but I think this is because the test is much simpler and doesn't do as much of the other work the real application does. It does demonstrate the problem, even if more pronounced than the real application.

Compiled and run on Kubuntu 18.10, gcc 8.2, x86_64, Core i7-4700HQ. TBB 2019 Update 2, also reproduces with Update 1.
[tbb_cpu_overhead_test.tar.gz](https://github.com/01org/tbb/files/2635928/tbb_cpu_overhead_test.tar.gz)

**Comment by Lastique** [comment]
I'm attaching the performance report from operf. You can see that the majority of time is spent in the kernel. Not sure what causes the `fgetpwent` entry, the test doesn't call it, neither does TBB, AFAICT. 

The top TBB symbols are various mutex methods. I'm not considering construct/destroy methods as there don't cause context switches. The important ones I consider the `queuing_rw_mutex::scoped_lock` methods, `concurrent_vector_base_v3::helper::enable_segment`, `pipeline_root_task::execute` and probably others, where I can see a lot of spinning.

[perf.txt](https://github.com/01org/tbb/files/2635948/perf.txt)

**Comment by Lastique** [comment]
I've created https://github.com/01org/tbb/pull/106 that significantly reduces the CPU overhead, although it causes test_task_arena failure, as described in https://github.com/01org/tbb/issues/104. I don't know how critical that failure is.

I don't consider that PR a final solution for the problem. Although it improves the situation, I'm not sure it will be as beneficial for every kind of workload. And I think the situation can be further improved if TBB is updated to employ blocking, when spinning is inefficient.

**Comment by akukanov** [comment]
Thank you for a very detailed problem description and the reproducer. We will investigate.

**Comment by Lastique** [comment]
Any updates?

**Comment by Lastique** [comment]
Ping?

**Comment by Lastique** [comment]
Guys, it's been a year. @akukanov were you able to investigate?

**Comment by alexey-katranov** [comment]
@Lastique, I looked through the example, it seems relatively stressful for TBB scheduler because there are 256 implicit arenas that are waken up with parallel activity every 20ms/256 = 80 us. It is comparable to the time when worker threads decide to go to sleep. In this case with so many arenas, it is possible that threads never decide to sleep (due to overheads of thread migration between arenas).
Can your approach be modified to use the only one arena (e.g. explicitly created with `task_arena`) to proceed all IO requests?

**Comment by Lastique** [comment]
I'm not sure how I can do this. I haven't used `task_arena` before and from the description it is not clear how to use it with `parallel_for`.

Note that:

- I cannot reduce the number of IO threads or change the frequency of arrival of work bundles.
- Merging work bundles is highly undesirable since the work must be performed in soft realtime.
- Because of the soft realtime requirement (and CPU overhead) passing work bundles for processing to another thread (which might utilize TBB) is also not really an option.

Basically, I need to process data in IO threads, and offload some of the work to other TBB threads, if possible, to reduce processing time. I.e. the TBB threads are supposed to be mere helpers and impose as little overhead as possible to allow IO threads to proceed.

**Comment by alexey-katranov** [comment]
`arena` is a work sharing primitive where threads can find and proceed work. Each thread can proceed work only inside the current `arena`, i.e. worker threads cannot steal work between arenas. By default, TBB creates an implicit `arena` for each user thread that works with TBB. In the case of 256 external threads it might be suboptimal because it involves some overhead of worker thread balancing between `arena`s.
[`task_arena`](https://spec.oneapi.com/versions/latest/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.html) is a public API to create user manageable `arena`s. In your case, I would try to create only one `task_arena` for all parallel_work requests:
```
int max_threads = tbb::this_task_arena::max_concurrency() 
        + 256 /* the number of IO threads to reserve slots inside the arena */;
tbb::task_arena arena(max_threads, 256);
```
To submit work to the `task_arena` you can use `execute` (blocks until work is done) or `enqueue` (does not block the calling thread). In your case, I believe you can use `execute`:
```
arena.execute([] {
    tbb::parallel_for(...);
});
```
I would recommend to use isolation to prevent an IO thread to participate in parallel work submitted by another IO thread:
```
arena.execute([] {
    tbb::this_task_arena::isolate([] {
        tbb::parallel_for(...);
    });
});
```
In addition, small optimization is possible if the IO thread joins (via `execute`) the `task_arena` only once:
```
void io_loop() {
    arena.execute([] {
        tbb::this_task_arena::isolate([] {
            while (!m_stop_flag.load(std::memory_order_relaxed)) {
                // Input data comes in chunks, roughly every 20 ms
                std::this_thread::sleep_for(std::chrono::milliseconds(20));
                ...
                tbb::parallel_for(...);
                ...
        });
    });
}
```

**Comment by Lastique** [comment]
The problem is that I don't have a fixed number of IO threads, and their number is also not bounded by a pre-defined limit. If I understand correctly, if the number of threads calling `execute` exceeds the limit specified in the `task_arena` constructor, then some of the threads must block, which is not acceptable. In my case, consumers may block in their own right, and may also take considerable time to process their work bundle, which then makes it possible for the producers to also block in the arena.

Is it possible to keep all IO threads running while making all of them use a single task arena, from which TBB worker threads will pick tasks when possible? If no free TBB threads are available (e.g. when they are blocked or descheduled by kernel) for the tasks generated by a given IO thread then at least the IO thread must proceed with these tasks.
